subgraphs/rari-yag-v1/src/mappings/helpers.ts

```diff
@@ -0,0 +1,389 @@
+  afterFeeUSD: BigInt, // used to calculate withdrawal fee - only used in Yield Pool
+  outputTokensBurned: BigInt,
+  asset: string,
+  vaultAddress: string,
+): void {
+  // create id
+  let hash = event.transaction.hash;
+  let logIndex = event.transaction.index;
+  let id = "withdraw-" + hash.toHexString() + "-" + logIndex.toString();
+
+  // create Deposit
+  let withdraw = new Withdraw(id);
+
+  // get (or create) token
+  let token = getOrCreateToken(asset);
+  let vault = getOrCreateVault(event, vaultAddress, token.id);
+  withdraw.vault = vault.id;
+
+  // populate vars,
+  withdraw.hash = hash.toHexString();
+  withdraw.logIndex = logIndex.toI32();
+  withdraw.protocol = RARI_DEPLOYER;
+  withdraw.to = event.transaction.from.toHexString();
+  withdraw.from = event.address.toHexString();
+  withdraw.blockNumber = event.block.number;
+  withdraw.timestamp = event.block.timestamp;
+  withdraw.asset = token.id;
+  withdraw.amount = amount;
+  withdraw.amountUSD = amountUSD;
+
+  withdraw.save();
+
+  updateTVL(event);
+
+  // update outputTokenPrice
+  vault.outputTokenSupply = vault.outputTokenSupply!.minus(outputTokensBurned);
+  vault.outputTokenPriceUSD = getOutputTokenPrice(vault.outputTokenSupply!, vaultAddress);
+
+  // calculate withdrawal fee
+  let withdrawalFee = BIGDECIMAL_ZERO;
+  if (vaultAddress == YIELD_VAULT_ADDRESS) {
+    withdrawalFee = amountUSD.minus(afterFeeUSD.toBigDecimal().div(exponentToBigDecimal(DEFAULT_DECIMALS)));
+    if (withdrawalFee.lt(BIGDECIMAL_ZERO)) {
+      withdrawalFee = BIGDECIMAL_ZERO;
+    }
+  }
+
+  // update fees/revenues/token balances
+  updateYieldFees(vaultAddress);
+  updateRevenues(event, vault, withdrawalFee);
+  updateInputTokenBalance(vault);
+
+  // calculate pricePerShare
+  vault.pricePerShare = vault.inputTokenBalance
+    .toBigDecimal()
+    .div(exponentToBigDecimal(DEFAULT_DECIMALS))
+    .div(vault.outputTokenSupply!.toBigDecimal().div(exponentToBigDecimal(DEFAULT_DECIMALS)));
+
+  vault.save();
+}
+
+/////////////////////////
+//// Updates Helpers ////
+/////////////////////////
+
+// updates yield fees if necessary
+export function updateYieldFees(vaultAddress: string): void {
+  let perfFee: VaultFee = getOrCreateVaultFee(VaultFeeType.PERFORMANCE_FEE, vaultAddress);
+  let tryWithdrawalFee: ethereum.CallResult<BigInt>;
+  let tryPerfFee: ethereum.CallResult<BigInt>;
+
+  if (vaultAddress == YIELD_VAULT_ADDRESS) {
+    // get fees
+    let contract = RariYieldFundManager.bind(Address.fromString(YIELD_VAULT_MANAGER_ADDRESS));
+    let withdrawalFee = getOrCreateVaultFee(VaultFeeType.WITHDRAWAL_FEE, vaultAddress);
+    tryWithdrawalFee = contract.try_getWithdrawalFeeRate();
+    tryPerfFee = contract.try_getInterestFeeRate();
+
+    if (!tryWithdrawalFee.reverted) {
+      // only update fee if it was not reverted
+      withdrawalFee.feePercentage = tryWithdrawalFee.value
+        .toBigDecimal()
+        .div(exponentToBigDecimal(DEFAULT_DECIMALS - 2));
```

Comment:
Hello @dmelotik , can you avoid hard coding `DEFAULT_DECIMALS`, you should use the `inputToken.decimals` instead.

Reply:
Good catch I did change some of these to `inputToken.decimals`

However there are some places where the BigInt is offset by 1e18 on the contract-level. And here is the contract call for this one: https://github.com/Rari-Capital/rari-yield-pool-contracts/blob/master/contracts/RariFundManager.sol#L954